### PR TITLE
Use DenseLayout in optimization_level=1 if not swap mapped.

### LINKED ISSUE #2845: optimization_level=1 always uses Trivial layout
### Information

- **Qiskit Terra version**: master
- **Python version**:
- **Operating system**:

### What is the current behavior?
```
               ┌───┐                ░ ┌─┐            
q37_0: |0>─────┤ X ├────────────────░─┤M├────────────
          ┌───┐└─┬─┘┌───┐           ░ └╥┘┌─┐         
q37_1: |0>┤ H ├──■──┤ X ├───────────░──╫─┤M├─────────
          ├───┤     └─┬─┘┌───┐      ░  ║ └╥┘┌─┐      
q37_2: |0>┤ H ├───────■──┤ X ├──────░──╫──╫─┤M├──────
          ├───┤          └─┬─┘┌───┐ ░  ║  ║ └╥┘┌─┐   
q37_3: |0>┤ H ├────────────■──┤ X ├─░──╫──╫──╫─┤M├───
          ├───┤               └─┬─┘ ░  ║  ║  ║ └╥┘┌─┐
q37_4: |0>┤ H ├─────────────────■───░──╫──╫──╫──╫─┤M├
          └───┘                     ░  ║  ║  ║  ║ └╥┘
  c5_0: 0 ═════════════════════════════╩══╬══╬══╬══╬═
                                          ║  ║  ║  ║ 
  c5_1: 0 ════════════════════════════════╩══╬══╬══╬═
                                             ║  ║  ║ 
  c5_2: 0 ═══════════════════════════════════╩══╬══╬═
                                                ║  ║ 
  c5_3: 0 ══════════════════════════════════════╩══╬═
                                                   ║ 
  c5_4: 0 ═════════════════════════════════════════╩═
```
```python
backend = provider.get_backend('ibmq_20_tokyo')
new_qc = transpile(qc, backend, optimization_level=1)
```

shows initial layout as the trivial layout, which is disconnected:

Screen Shot 2019-07-22 at 4 14 54 PM

However part #2 in the source says `_layout_check` does better layout 
if the circuit needs swaps, 

```python
# 2. Use a better layout on densely connected qubits, if circuit needs swaps
```
which is obviously the case here, but no other layouts
are actually selected (or imported).

### What is the expected behavior?
Expected that the `DenseLayout` is used if there are swaps needed.

### PR DESCRIPTION
**Pull request by nonhermitian**
### Summary
Fixes #2845.

### CODE REVIEW

**Comment by 1ucian0** [comment]
I added tests. Maybe level 2 also needs a look.

**Comment by 1ucian0** [comment]
This also fixes #2260, @kdk ?

**Comment by kdk** [comment]
This partially addresses #2260. We're still not running `_direction_check` anywhere, and we should be sure to fix this consistently across the optimization levels.

File: test/python/transpiler/test_preset_passmanagers.py [hunk 1]
```diff
@@ -172,3 +172,50 @@ def test_layout_2503(self, level):
...
+                                  10: ancilla[5], 11: ancilla[6], 12: ancilla[7], 13: ancilla[8],
+                                  14: ancilla[9], 15: ancilla[10], 16: ancilla[11], 17: ancilla[12],
+                                  18: ancilla[13], 19: ancilla[14]}
+        expected_layouts = [expected_layout_level0,
+                            expected_layout_level1,
+                            expected_layout_level2,
+                            expected_layout_level3]
+        backend = FakeTokyo()
+        result = transpile(qc, backend, optimization_level=level, seed_transpiler=42)
+        self.assertEqual(result.layout._p2v, expected_layouts[level])
```
Comment (kdk) [inline test/python/transpiler/test_preset_passmanagers.py]: Can we add some detail around the meaning of the results? It looks like here, that we're testing that for under optimization_levels  0, 2, and 3, we're getting the trivial layout, and for 1 the dense layout, but it's not immediately clear.

Maybe we could have two tests? One to verify that  circuit that that would be mapped under the trivial layout is assigned the trivial layout (across optimization_levels), and another to verify that when the circuit would not be mapped under the trivial layout, the DenseLayout is used.

Comment (kdk) [reply test/python/transpiler/test_preset_passmanagers.py]: Also, the tests should include the case where the input circuit is defined across two quantum registers.

**Comment by 1ucian0** [comment]
> This partially addresses #2260. We're still not running `_direction_check` anywhere, and we should be sure to fix this consistently across the optimization levels.

Plus PR #2883 ?

**Comment by kdk** [comment]
:+1: #2883 covers the rest of #2260

**Comment by 1ucian0** [comment]
This PR was replaced by PR #2971

**Pull request closed without merging**